test/gtest/ucp/test_ucp_peer_failure.cc

```diff
@@ -570,3 +621,160 @@ UCS_TEST_P(test_ucp_peer_failure_keepalive, kill_receiver,
+        set_am_handler(receiver());
+
+        m_sbuf.resize(UCS_KBYTE);
+        m_rbuf.resize(UCS_KBYTE);
+    }
+
+    static void get_test_variants(std::vector<ucp_test_variant> &variants)
+    {
+        add_variant_with_value(variants, UCP_FEATURE_AM, TEST_AM, "am");
+        add_variant_with_value(variants, UCP_FEATURE_TAG, TEST_TAG, "tag");
+    }
+
+    template<typename Func>
+    void for_each_protos_progress(Func func)
+    {
+        for (int proto_id = 0; proto_id < ucp_protocols_count(); ++proto_id) {
+            auto proto = const_cast<ucp_proto_t*>(ucp_protocols[proto_id]);
+            int stage = UCP_PROTO_STAGE_START;
+            for (; stage < UCP_PROTO_STAGE_LAST; ++stage) {
+                func(proto, stage);
+            }
+        }
+    }
+
+    void protos_replace_progress()
+    {
+        auto replace_progress = [&](ucp_proto_t* proto, int stage) {
+            m_progress_storage[proto][stage] = proto->progress[stage];
+            proto->progress[stage]           = &progress_wrapper;
+        };
+        for_each_protos_progress(replace_progress);
+    }
+
+    void protos_restore_progress()
+    {
+        auto restore_progress = [&](ucp_proto_t* proto, int stage) {
+            proto->progress[stage] = m_progress_storage[proto][stage];
+        };
+        for_each_protos_progress(restore_progress);
+    }
+
+    void replace_ep_put_zcopy(ucp_ep_h ep)
+    {
+        for (auto lane = 0; lane < ucp_ep_num_lanes(ep); ++lane) {
+            if (lane == ucp_ep_get_cm_lane(ep)) {
+                continue;
+            }
+
+            auto iface  = ucp_ep_get_lane(ep, lane)->iface;
+            auto result = m_sender_uct_ops.emplace(iface, iface->ops);
+            if (result.second) {
+                iface->ops.ep_put_zcopy =
+                        reinterpret_cast<uct_ep_put_zcopy_func_t>(
+                                ucs_empty_function_return_unsupported);
+            }
+        }
+    }
+
+    static ucs_status_t progress_wrapper(uct_pending_req_t *uct_req)
+    {
+        auto req        = ucs_container_of(uct_req, ucp_request_t, send.uct);
+        auto proto      = req->send.proto_config->proto;
+        auto stage      = req->send.proto_stage;
+        auto proto_name = req->send.proto_config->proto->name;
+
+        if (std::string(proto_name) == "rndv/put/zcopy") {
+            if (stage == UCP_PROTO_STAGE_START) {
+                if (!self->m_is_ep_closed) {
+                    scoped_log_handler slh(wrap_errors_logger);
+                    self->receiver().close_all_eps(*self, 0,
+                                                   UCP_EP_CLOSE_FLAG_FORCE);
+                    self->replace_ep_put_zcopy(req->send.ep);
+                    self->m_is_ep_closed = true;
+                }
+            } else {
+                EXPECT_TRUE(false) << "Unexpected proto stage";
+            }
+        }
+
+        ucs_status_t status = self->m_progress_storage[proto][stage](uct_req);
```

Comment:
can we safe this pointer in the request instead of global var?

Reply:
I checked it and it doesn't work. `user_data` for send requests is set after the `ucp_request_send()` (and therefore `progress()` call.

Reply:
ok